server/src/main/java/org/elasticsearch/search/profile/SearchProfileResults.java

```diff
@@ -170,4 +196,53 @@ private static void parseProfileResultsEntry(XContentParser parser, Map<String,
         result.getQueryPhase().setSearchProfileDfsPhaseResult(searchProfileDfsPhaseResult);
         searchProfileResults.put(id, result);
     }
+
+    /**
+     * Parsed representation of a composite id used for shards in a profile.
+     * The composite id format is specified/created via the {@code SearchShardTarget} method.
+     * @param nodeId nodeId that the shard is on
+     * @param indexName index being profiled
+     * @param shardId shard id being profiled
+     * @param clusterName if a CCS search, the remote clusters will have a name in the id. Local clusters will be null.
+     */
+    record ShardProfileId(String nodeId, String indexName, int shardId, @Nullable String clusterName) {}
+
+    /**
+     * Parse the composite "shard id" from the profiles output, which comes from the
+     * {@code SearchShardTarget.toString()} method, into its separate components.
+     * <p>
+     * One of two expected patterns is accepted:
+     * <p>
+     * 1) [nodeId][indexName][shardId]
+     * example: [2m7SW9oIRrirdrwirM1mwQ][blogs][1]
+     * <p>
+     * 2) [nodeId][clusterName:indexName][shardId]
+     * example: [UngEVXTBQL-7w5j_tftGAQ][remote1:blogs][0]
+     *
+     * @param compositeId see above for accepted formats
+     * @return ShardProfileId with parsed components or null if the compositeId has an unsupported format
+     */
+    static ShardProfileId parseCompositeProfileShardId(String compositeId) {
+        assert Strings.isNullOrEmpty(compositeId) == false : "An empty id should not be passed to parseCompositeProfileShardId";
+
+        Pattern r = Pattern.compile("\\[([^]]+)\\]\\[([^]]+)\\]\\[(\\d+)\\]");
```

Comment:
Can we make this a `private static final Pattern SHARD_ID_DECOMPOSITION = Pattern.compile("\\[([^]]+)\\]\\[([^]]+)\\]\\[(\\d+)\\]");`

For every profile result, we could have many different IDs and while profiling is already slow as it bypasses cache and adds additional overhead, lets not needlessly create short lived objects and recompile a regex pattern.

Reply:
Fixed.